### PR TITLE
Feature/admin panel

### PR DESCRIPTION
**Pull request by wolftender**
added basic user admin panel, admins (users with `userStatus == 1`) can now:

- ban users
- unban users
- edit logins

### CODE REVIEW

File: web/src/api/ApiClient.ts [hunk 1]
```diff
@@ -76,6 +76,18 @@ export class ApiClient {
         }
     }
 
+    async getUsers(): Promise<User[] | null> {
+        const users: any = await this.baseRequest<User>(`user`, {
```
Comment (shilangyu) [inline web/src/api/ApiClient.ts:80]: `any` not needed

Comment (shilangyu) [reply web/src/api/ApiClient.ts:80]: Should be `baseRequest<User[]>`

File: web/src/api/ApiClient.ts [hunk 2]
```diff
@@ -76,6 +76,18 @@ export class ApiClient {
...
+            method: 'GET',
+        })
+
+        if (users instanceof Array) {
+            return users
+        }
+
+        return null
```
Comment (shilangyu) [inline web/src/api/ApiClient.ts:88]: It is never null

File: web/src/pages/AdminPage/UserEditor.tsx [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    DialogContentText,
+    DialogTitle,
+    TextField,
+} from '@mui/material'
+import * as React from 'react'
+import { useEffect, useState } from 'react'
+import { User, UserUpdateRequest } from '../../model/user'
+
+interface UserEditorProps {
+    user: User | null
```
Comment (shilangyu) [inline web/src/pages/AdminPage/UserEditor.tsx:15]: In JS `undefined` tends to be used over `null` (easier to work with). It just a nitpick tho, feel free to ignore.

You could then just:
```ts
 user?: User
 ```

File: web/src/pages/AdminPage/UserEditor.tsx [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    onSave: (userId: number, newUser: UserUpdateRequest) => Promise<void>
+    onCancel: () => void
+}
+
+const UserEditor: React.FC<UserEditorProps> = (props: UserEditorProps) => {
+    const [user, setUser] = useState<User | null>(null)
```
Comment (shilangyu) [inline web/src/pages/AdminPage/UserEditor.tsx:21]: This does not have to be a hook, you can just use user from the props

File: web/src/pages/AdminPage/UserEditor.tsx [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+import { useEffect, useState } from 'react'
+import { User, UserUpdateRequest } from '../../model/user'
+
+interface UserEditorProps {
+    user: User | null
+    onSave: (userId: number, newUser: UserUpdateRequest) => Promise<void>
+    onCancel: () => void
+}
+
+const UserEditor: React.FC<UserEditorProps> = (props: UserEditorProps) => {
```
Comment (shilangyu) [inline web/src/pages/AdminPage/UserEditor.tsx:20]: Type annotation on `props: UserEditorProps` not needed, already specified in `React.FC`

File: web/src/pages/AdminPage/UserEditor.tsx [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+    const [user, setUser] = useState<User | null>(null)
+    const [error, setError] = useState<string | null>(null)
```
Comment (shilangyu) [inline web/src/pages/AdminPage/UserEditor.tsx:22]: Never used. Maybe just display a snackbar?

File: web/src/pages/AdminPage/UserEditor.tsx (hunk 3 above)
Comment (shilangyu) [inline web/src/pages/AdminPage/UserEditor.tsx:15]: Why is it nullable? When do we want to edit a non-existing user?

File: web/src/pages/AdminPage/index.tsx [hunk 7]
```diff
@@ -0,0 +1,65 @@
+import SettingsIcon from '@mui/icons-material/Settings'
+import { Card, Container, Grid, Typography, useTheme } from '@mui/material'
+import * as React from 'react'
+import UserList from './UserList'
+
+function AdminPage() {
+    const theme = useTheme()
+
+    return (
+        <>
```
Comment (shilangyu) [inline web/src/pages/AdminPage/index.tsx:10]: `Fragment` not needed

File: web/src/pages/AdminPage/UserList.tsx [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+            .then(() => {
+                refreshUserList()
+            })
+            .catch(err => {
+                enqueueSnackbar(`failed to delete user ${userId}`, {
+                    variant: 'error',
+                })
+                console.error(`failed to delete user: ${err}`)
+            })
+    }
```
Comment (shilangyu) [inline web/src/pages/AdminPage/UserList.tsx:108]: Why not `async`/`await`?

File: web/src/pages/AdminPage/UserList.tsx [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+import { useEffect, useState } from 'react'
+import { apiClient } from '../../api'
+import { User, UserUpdateRequest } from '../../model/user'
+import UserEditor from './UserEditor'
+
+const UserList: React.FC<Record<string, never>> = (
+    props: Record<string, never>
+) => {
+    const [users, setUsers] = useState<User[] | null>(null)
+    const [error, setError] = useState<string | null>(null)
```
Comment (shilangyu) [inline web/src/pages/AdminPage/UserList.tsx:22]: Never used

File: web/src/pages/AdminPage/UserList.tsx [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+    Typography,
+} from '@mui/material'
+import { useSnackbar } from 'notistack'
+import * as React from 'react'
+import { useEffect, useState } from 'react'
+import { apiClient } from '../../api'
+import { User, UserUpdateRequest } from '../../model/user'
+import UserEditor from './UserEditor'
+
+const UserList: React.FC<Record<string, never>> = (
```
Comment (shilangyu) [inline web/src/pages/AdminPage/UserList.tsx:18]: This has no props, just leave empty

**Review by shilangyu — CHANGES_REQUESTED** [review]

File: web/src/pages/AdminPage/UserEditor.tsx (hunk 3 above)
Comment (wolftender) [reply web/src/pages/AdminPage/UserEditor.tsx:15]: i left it like that to later add a button like `add user` at the bottom of the list, i can remove it if you want

File: web/src/pages/AdminPage/UserList.tsx (hunk 8 above)
Comment (wolftender) [reply web/src/pages/AdminPage/UserList.tsx:108]: no reason

**Review by Zaczero — APPROVED** [review]

**Review by shilangyu — APPROVED** [review]

**Pull request merged**